### PR TITLE
Adding support for replace in selection (#234)

### PR DESCRIPTION
**Pull request by AndersonDaniel**
Hi there!

Addition of the replace "in selection" functionality in notepad++.

### CODE REVIEW

**Comment by danieleds** [comment]
Hi, thank you. 
If I am not mistaken, this only works with the "Replace All" button. I think we need to add support for search in selection too, otherwise the feature is confusing.

**Comment by AndersonDaniel** [comment]
Yes, this currently only works with the "Replace All" button.
You mean that it should also work with the "Search All" button, and the effect would be to shrink the current selection to the search term?
For example, given the following text and selection (selection in bold):

> Lorem ipsum dolor sit amet
> **Lorem ipsum dolor sit amet**
> Lorem ipsum dolor sit amet

and the search term "lor", after clicking "Search All" with "in selection" checked the selected text will be:

> Lorem ipsum dolor sit amet
> **Lor**em ipsum do**lor** sit amet
> Lorem ipsum dolor sit amet

**Comment by JuBan1** [comment]
Yes, I'm pretty sure that's wat he meant. 

Would be great if it worked the same way Replace works. Meaning the "Search In Selection" is automatically checked if the user opens the search dialog while having some text selected.

**Comment by danieleds** [comment]
Yes, it's like @JuBan1 said.

And it should also work with "Search prev / search next / replace prev / replace next"

**Comment by AndersonDaniel** [comment]
Okay, so:

1. Implemented the "Search all in selection" functionality
2. Differentiating between two possibilities when the user searches and there is a selection:
    a) Multi-line selection - "In Selection" checkbox is automatically checked, selection isn't copied to search text
    b) Single line selection - "In Selection" is unchecked and selection is copied to search text.
3. While "In Selection" is checked, editing the search text (in "Find" tab) doesn't trigger the "Find next" operation (otherwise the selection is lost and there's no meaning to the "search all in selection" operation)

Concerning search next / prev, replace next / prev - any of them replaces the current selection, so having "in selection" affect them means the operation will only work once (after which the selection will be the first search /replace result). Seems to me a bit strange behavior... What do you think?

**Comment by JuBan1** [comment]
'Replace All' works great. "Select All" still selects *all* instances of the searched term for me.

> Concerning search next / prev, replace next / prev - any of them changes the current selection, so having "in selection" affect them means the operation will only work once (after which the selection will be the first search /replace result). Seems to me a bit strange behavior... What do you think?

Yes, that's a bit of a problem. If it turns out Find Next/Prev don't work then they should probably be greyed out if 'In Selection' is checked.

Simply remembering the start and end of the selection and still allowing Find Next/Prev is probably a bad idea since it would be confusing. Users should be able to clearly see the part of the text that is searched. 

A possible solution is to use some highlight color (similar to what happens when you select a word and all other occurrences of this word are highlighted as well). When the user uses Find Next/Prev the selection could be used as usual (to show the found occurrence) and the search area could be highlighted with a different color. 

On the downside, it's probably a little more involved to implement.

**Comment by AndersonDaniel** [comment]
Added the part of graying out the Next/Prev buttons when 'In Selection' is active.

About "Select All" still selecting all instances, it could be a bit confusing since when even a single instance is selected, all of them are highlighted (not only in selection). For me it works, could you please verify this is not the case? (For example by using "Select All" with "In Selection" checked and then editing the selected text.)

Thanks!

**Comment by JuBan1** [comment]
Yes, you're right. I confused the highlighting with selection, derp.

**Comment by JuBan1** [comment]
Alright, looking good. I have two more suggestions.

First, In Selection should probably be toggleable when the search window is opened without any text selected. So users can turn on In Selection search without having to close the window, select some text, then reopen it. Right now the checkbox is disabled.

Next, I thought wrapping the search and replace buttons as well as the selection checkbox in a QFrame might be a nice way to visually group them. Here's a mock-up:



Although I don't know how good that would look with the way the search UI is constructed (= buttons being re-used and layout changed depending on which tab is selected).

**Comment by AndersonDaniel** [comment]
Fixed both, QFrame looks fine I think:
qframe_find
qframe_replace

What do you think?

File: src/ui/include/mainwindow.h [hunk 1]
```diff
@@ -293,6 +293,7 @@ private slots:
     QString             currentWordOrSelection();
     void                currentWordOnlineSearch(const QString &searchUrl);
     QString             getNewDocumentName();
+    void                showSearchReplace(frmSearchReplace::Tabs tab);
```
Comment (JuBan1) [inline src/ui/include/mainwindow.h]: One last round of nagging and I'll be content.

Could you add a short documentation to this new function?

File: src/ui/Search/frmsearchreplace.ui [hunk 2]
```diff
@@ -324,15 +315,44 @@
...
-       </widget>
+      <item row="0" column="3" rowspan="3">
+        <widget class="QFrame" name="frameSearchReplaceAll">
+          <property name="frameShape">
+            <enum>QFrame::StyledPanel</enum>
+          </property>
+          <property name="frameShadow">
+            <enum>QFrame::Plain</enum>
+          </property>
+          <layout class="QVBoxLayout" name="verticalLayout_3">
```
Comment (JuBan1) [inline src/ui/Search/frmsearchreplace.ui]: I'm getting the following warning:

> Search/frmsearchreplace.ui: Warning: The name 'verticalLayout_3' (QVBoxLayout) is already in use, defaulting to 'verticalLayout_31'.

Name's already in use in line 96.

**Comment by JuBan1** [comment]
I like it. I tested it on my end as well and it feels pretty good to use.

**Comment by danieleds** [comment]
This is how Qt Creator does it:

ex

I like it a lot, but I'm not sure how easy it is to implement.
If it's not, then the current behavior is good but I'd like to think about the UI before merging.

**Comment by AndersonDaniel** [comment]
Fixed both things, thanks :)

**Comment by AndersonDaniel** [comment]
Started playing with CodeMirror highlighting. Surrounding the selection with a border like Qt Creator seems rather difficult; we could highlight the previously selected text something like this:
example
But I don't know if it looks intuitive enough, and I also think it would require a lot of code to handle special cases with replace when the replacement length is not equal to the replaced text's length (also for regexp replacing).

**Comment by JuBan1** [comment]
How does Replace In Selection handle it right now? Shouldn't it already have to handle the change of the selected code?

**Comment by AndersonDaniel** [comment]
Yes, CodeMirror internally takes care of that when replacing. From the docs:

> **replace**(text: string, ?origin: string)
> Replaces the currently found match with the given text and _adjusts the cursor position to reflect the replacement._

**Comment by Teklad** [comment]
If you accessed the drawing functions a bit, you could probably paint an overlay using QPainter, but I debate how hacky that would be, lol.

Anyways, this may have a few conflicts with #263 in terms of merging, so it'd probably be better if this pull request was merged first.  The changes in this request are much smaller so I shouldn't have any problems fixing conflicts when rebasing if I run into any.

**Comment by valera-rozuvan** [comment]
@AndersonDaniel Any progress?

**Comment by AndersonDaniel** [comment]
@valera-rozuvan I didn't pursue the previous-selection-highlighting direction further (I think it would be too complicated to do it properly). The rest was ready for merge since the last commit, I see conflicts have accumulated since then.

**Comment by valera-rozuvan** [comment]
I am guessing this PR is no longer desirable, and too far behind the `master` branch? Maybe close it?

@danieleds , @AndersonDaniel , @JuBan1 - Your thoughts?

**Pull request closed without merging**